File: solutions/php/annalyns-infiltration/1/AnnalynsInfiltration.php
<?php

class AnnalynsInfiltration
{
    public function canFastAttack($is_knight_awake)
    {
        // Fast attack can be made if the knight is sleeping
        return !$is_knight_awake;
    }

    public function canSpy($is_knight_awake, $is_archer_awake, $is_prisoner_awake)
    {
        // The group can be spied upon if at least one of them is awake
        return $is_knight_awake || $is_archer_awake || $is_prisoner_awake;
    }

    public function canSignal($is_archer_awake, $is_prisoner_awake)
    {
        // The prisoner can be signalled if the prisoner is awake and the archer is sleeping
        return $is_prisoner_awake && !$is_archer_awake;
    }

    public function canLiberate($is_knight_awake, $is_archer_awake, $is_prisoner_awake, $is_dog_present)
    {
        if ($is_dog_present) {
            // If Annalyn has her dog, she can rescue if the archer is asleep
            return !$is_archer_awake;
        } else {
            // If Annalyn doesn't have her dog, she can free the prisoner if:
            // - The prisoner is awake
            // - The knight and archer are both sleeping
            return $is_prisoner_awake && !$is_knight_awake && !$is_archer_awake;
        }
    }
}

File: solutions/r/annalyns-infiltration/1/annalyns-infiltration.R
can_fast_attack <- function(knight_is_awake) {
  # Fast attack is possible if the knight is NOT awake
  !knight_is_awake
}

can_spy <- function(knight_is_awake, archer_is_awake, prisoner_is_awake) {
  # Spying is possible if at least one character is awake
  knight_is_awake || archer_is_awake || prisoner_is_awake
}

can_signal_prisoner <- function(archer_is_awake, prisoner_is_awake) {
  # Signaling is possible if prisoner is awake AND archer is NOT awake
  prisoner_is_awake && !archer_is_awake
}

can_free_prisoner <- function(knight_is_awake,
                              archer_is_awake,
                              prisoner_is_awake,
                              pet_dog_is_present) {
  # Two ways to free the prisoner:
  # 1. With dog: archer must be asleep (knight is scared of dog)
  # 2. Without dog: prisoner awake AND knight asleep AND archer asleep
  (pet_dog_is_present && !archer_is_awake) ||
    (!pet_dog_is_present && prisoner_is_awake && !knight_is_awake && !archer_is_awake)
}